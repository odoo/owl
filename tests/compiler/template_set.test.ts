import { snapshotEverything, TestContext } from "../helpers";

snapshotEverything();

describe("loading templates", () => {
  test("can initialize qweb with a string", () => {
    const templates = `<?xml version="1.0" encoding="UTF-8"?>
      <templates id="template" xml:space="preserve">
        <div t-name="hey">jupiler</div>
      </templates>`;
    const context = new TestContext();
    context.addTemplates(templates);
    expect(context.renderToString("hey")).toBe("<div>jupiler</div>");
  });

  test("can initialize qweb with an XMLDocument", () => {
    const data = `<?xml version="1.0" encoding="UTF-8"?>
      <templates id="template" xml:space="preserve">
        <div t-name="hey">jupiler</div>
      </templates>`;
    const xml = new DOMParser().parseFromString(data, "text/xml");
    const context = new TestContext();
    context.addTemplates(xml);
    expect(context.renderToString("hey")).toBe("<div>jupiler</div>");
  });

  test("addTemplates does not modify its xml document in place", () => {
    const data = `<?xml version="1.0" encoding="UTF-8"?>
      <templates id="template" xml:space="preserve"><div t-name="hey"><t t-esc="value"/></div></templates>`;
    const xml = new DOMParser().parseFromString(data, "text/xml");
    const context = new TestContext();
    expect(xml.firstElementChild!.innerHTML).toBe(`<div t-name="hey"><t t-esc="value"/></div>`);
    context.addTemplates(xml);
    expect(context.renderToString("hey", { value: 123 })).toBe("<div>123</div>");
    expect(xml.firstElementChild!.innerHTML).toBe(`<div t-name="hey"><t t-esc="value"/></div>`);
  });

  test("can load a few templates from a xml string", () => {
    const data = `<?xml version="1.0" encoding="UTF-8"?>
      <templates id="template" xml:space="preserve">

        <t t-name="items"><li>ok</li><li>foo</li></t>

        <ul t-name="main"><t t-call="items"/></ul>
      </templates>`;
    const context = new TestContext();
    context.addTemplates(data);
    const result = context.renderToString("main");
    expect(result).toBe("<ul><li>ok</li><li>foo</li></ul>");
  });

  test("can load a few templates from an XMLDocument", () => {
    const data = `<?xml version="1.0" encoding="UTF-8"?>
      <templates id="template" xml:space="preserve">

        <t t-name="items"><li>ok</li><li>foo</li></t>

        <ul t-name="main"><t t-call="items"/></ul>
      </templates>`;
    const xml = new DOMParser().parseFromString(data, "text/xml");
    const context = new TestContext();
    context.addTemplates(xml);
    const result = context.renderToString("main");
    expect(result).toBe("<ul><li>ok</li><li>foo</li></ul>");
  });

  test("does not crash if string does not have templates", () => {
    const data = "";
    const context = new TestContext();
    context.addTemplates(data);
    expect(Object.keys(context.rawTemplates)).toEqual([]);
  });

  test("does not crash if XMLDocument does not have templates", () => {
    const data = "";
    const xml = new DOMParser().parseFromString(data, "text/xml");
    const context = new TestContext();
    context.addTemplates(xml);
    expect(Object.keys(context.rawTemplates)).toEqual([]);
  });
});
