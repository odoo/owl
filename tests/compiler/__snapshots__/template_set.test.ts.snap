// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loading templates can initialize qweb with a string 1`] = `
"function anonymous(bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, component, comment } = bdom;
  
  let block1 = createBlock(\`<div>jupiler</div>\`);
  
  return function template(ctx, node, key = \\"\\") {
    return block1();
  }
}"
`;

exports[`loading templates can initialize qweb with an XMLDocument 1`] = `
"function anonymous(bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, component, comment } = bdom;
  
  let block1 = createBlock(\`<div>jupiler</div>\`);
  
  return function template(ctx, node, key = \\"\\") {
    return block1();
  }
}"
`;

exports[`loading templates can load a few templates from a xml string 1`] = `
"function anonymous(bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, component, comment } = bdom;
  
  let block2 = createBlock(\`<li>ok</li>\`);
  let block3 = createBlock(\`<li>foo</li>\`);
  
  return function template(ctx, node, key = \\"\\") {
    let b2 = block2();
    let b3 = block3();
    return multi([b2, b3]);
  }
}"
`;

exports[`loading templates can load a few templates from a xml string 2`] = `
"function anonymous(bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, component, comment } = bdom;
  
  let block1 = createBlock(\`<ul/>\`);
  
  return function template(ctx, node, key = \\"\\") {
    return block1();
  }
}"
`;

exports[`loading templates can load a few templates from an XMLDocument 1`] = `
"function anonymous(bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, component, comment } = bdom;
  
  let block2 = createBlock(\`<li>ok</li>\`);
  let block3 = createBlock(\`<li>foo</li>\`);
  
  return function template(ctx, node, key = \\"\\") {
    let b2 = block2();
    let b3 = block3();
    return multi([b2, b3]);
  }
}"
`;

exports[`loading templates can load a few templates from an XMLDocument 2`] = `
"function anonymous(bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, component, comment } = bdom;
  
  let block1 = createBlock(\`<ul/>\`);
  
  return function template(ctx, node, key = \\"\\") {
    return block1();
  }
}"
`;
